File: src/PAdistData.f
* ====================================================================
*           R-ScaLAPACK version 0.4.x:  ScaLAPACK interface to R
*              Oak Ridge National Laboratory, Oak Ridge TN.
*        Authors: David Bauer, Guruprasad Kora, Nagiza. F. Samatova, 
*                            Srikanth Yoginath.
*     Contact: Nagiza F. Samatova; (865) 241-4351; samatovan@ornl.gov
*                 Computer Science and Mathematics Division
*             Oak Ridge National Laboratory, Oak Ridge TN 37831 
*                   (C) 2004 All Rights Reserved
*
*                              NOTICE
*
* Permission to use, copy, modify, and distribute this software and
* its documentation for any purpose and without fee is hereby granted
* provided that the above copyright notice appear in all copies and
* that both the copyright notice and this permission notice appear in
* supporting documentation.
*
* Neither the Oak Ridge National Laboratory nor the Authors make any
* representations about the suitability of this software for any
* purpose.  This software is provided ``as is'' without express or
* implied warranty.
*
* RScaLAPACK (http://www.aspect-sdm.org/Parallel-R) was funded
* as part of the Scientific Data Management Center
* (http://sdm.lbl.gov/sdmcenter) under the Department of Energy's 
* Scientific Discovery through Advanced Computing (DOE SciDAC) program
* (http://www.scidac.org ). 
* ======================================================================
*       Based on:
*
*       -- ScaLAPACK example code --
*       University of Tennessee, Knoxville, Oak Ridge National Laboratory,
*       and University of California, Berkeley.
*
*       Written by Antoine Petitet, August 1995 (petitet@cs.utk.edu)
*
* ==============================================================================

      SUBROUTINE PADistData(GlOBALMATRIX, PGINFO, WORKSIZ, MATNUM)
      DOUBLE PRECISION GLOBALMATRIX (*)
      INTEGER WORKSIZ
      INTEGER PGINFO (8)
      INTEGER MATNUM
*
*     Purpose
*     =======
*
*     PADistData distributes the input data matrix among different processes 
*     of the process grid as required by SCALAPACK functions
*
*  ================================================================================
*
*     .. Parameters ..
      INTEGER            BLOCK_CYCLIC_2D, CSRC_, CTXT_, DLEN_, DT_,
     $                   LLD_, MB_, M_, NB_, N_, RSRC_,KK
      PARAMETER          ( BLOCK_CYCLIC_2D = 1, DLEN_ = 9, DT_ = 1,
     $                     CTXT_ = 2, M_ = 3, N_ = 4, MB_ = 5, NB_ = 6,
     $                     RSRC_ = 7, CSRC_ = 8, LLD_ = 9 )
*     ..
*     .. Local Scalars ..
      INTEGER            H, I, IB, ICURCOL, ICURROW, II, J, JB,
     $                   JJ, K, M, N, NPCOL, NPROW, S2RANK
*     ..
*     .. Local Arrays ..
      DOUBLE PRECISION            WORK( WORKSIZ )
*     ..
*     .. External Subroutines ..
      EXTERNAL           BLACS_GRIDINFO, INFOG2L, DGERV2D, DGESD2D,
     $                   IGEBS2D, IGEBR2D
*     ..
*     .. Intrinsic Functions ..
      INTRINSIC          MAX,MIN
*     ..
*     .. Executable Statements ..
*
*     Get grid parameters
*

      IF (MATNUM.EQ.1) THEN
          M = PGINFO ( 1 )
          N = PGINFO ( 2 )
      ELSE
          M = PGINFO (3)
          N = PGINFO (4)
      ENDIF

       MB = PGINFO (5)
       NB = MB
       NPROW = PGINFO(7)
       NPCOL = PGINFO(8)


      KK = 1

      II = 1
      JJ = 1
      ICURROW = 0
      ICURCOL = 0
*
*     Loop over column blocks
*
*      IPA = 1
*      WORKSIZ = NB 
      
      DO 50 J = 1, N,  NB
         JB = MIN(  NB, N-J+1 )
         DO 40 H = 0, JB-1
*
*           Loop over block of rows
*
            DO 30 I = 1, M,  MB
               IB = MIN( MB , M-I+1 )
                     DO 20 K = 1, IB
                        WORK( K ) = GLOBALMATRIX ( KK )
*                         PRINT *, 'in pa_distdata -- ', GLOBALMATRIX(KK)
                        KK= KK + 1
   20                CONTINUE

                      S2RANK = ICURROW * NPCOL + ICURCOL
*                         PRINT *, 'Rank: ', S2RANK, ' data : ', WORK(1)
                      CALL PASendVectorToCR (IB,1,WORK,MB,S2RANK)
               ICURROW = MOD( ICURROW+1, NPROW )
   30       CONTINUE
*
            II = 1
            ICURROW = 0 
   40    CONTINUE
*
         ICURCOL = MOD( ICURCOL+1, NPCOL )
*
   50 CONTINUE
*
*
      RETURN
*
*     End of PADistData
*
      END

File: src/PAdistData.c
/* ====================================================================
*           R-ScaLAPACK version 0.4.x:  ScaLAPACK interface to R
*              Oak Ridge National Laboratory, Oak Ridge TN.
*        Authors: David Bauer, Guruprasad Kora, Nagiza. F. Samatova, 
*                            Srikanth Yoginath.
*     Contact: Nagiza F. Samatova; (865) 241-4351; samatovan@ornl.gov
*     Contact: Guruprasad Kora; (865) 576-6210; koragh@ornl.gov
*                 Computer Science and Mathematics Division
*             Oak Ridge National Laboratory, Oak Ridge TN 37831 
*                   (C) 2004 All Rights Reserved
*
*                              NOTICE
*
* Permission to use, copy, modify, and distribute this software and
* its documentation for any purpose and without fee is hereby granted
* provided that the above copyright notice appear in all copies and
* that both the copyright notice and this permission notice appear in
* supporting documentation.
*
* Neither the Oak Ridge National Laboratory nor the Authors make any
* representations about the suitability of this software for any
* purpose.  This software is provided ``as is'' without express or
* implied warranty.
*
* RScaLAPACK (http://www.aspect-sdm.org/Parallel-R) was funded
* as part of the Scientific Data Management Center
* (http://sdm.lbl.gov/sdmcenter) under the Department of Energy's 
* Scientific Discovery through Advanced Computing (DOE SciDAC) program
* (http://www.scidac.org ). 
* ======================================================================
*
*       Based on:
*
*       -- ScaLAPACK example code --
*       University of Tennessee, Knoxville, Oak Ridge National Laboratory,
*       and University of California, Berkeley.
*
*       Written by Antoine Petitet, August 1995 (petitet@cs.utk.edu)
*
* ==============================================================================
*/

#include "ParallelAgent.h"

void printContents (double *temp, int size)
{
	int i=0;
	for (i=0; i<size; i++)
		D_Rprintf(("vector [%d] - %f \n", i, *(temp+i)));
}

void PAdistData(double gMatrix[], int pgInfo[], int numRows, int numCols){

	int rowBlockSize, colBlockSize;
	int numProcessRows, numProcessCols;
	int currRow=0, currCol=0;
	int j=0;
	double *sendBuffer;
	int startIndex=0;
	int vectorSize;
	int one=1;
	int iCol=0, iRows=0;
	int sendToRank;
	int reqId=0;

	MPI_Request *requests;	
	MPI_Status *statuses;	
	rowBlockSize = pgInfo[4];
	colBlockSize = pgInfo[5];
	numProcessRows = pgInfo[6];
	numProcessCols = pgInfo[7];
/*
	used for non-blocking communication
	requests = (MPI_Request *) malloc (sizeof(MPI_Request) * rowBlockSize * iceil(numRows,rowBlockSize) * iceil(numCols,rowBlockSize));
	statuses = (MPI_Status *) malloc (sizeof(MPI_Status) * rowBlockSize * iceil(numRows,rowBlockSize) * iceil(numCols,rowBlockSize));
	
*/
	/* Loop over the column blocks */
	while (iCol < numCols)
	{	 
		j = 0;

		while (j < min( colBlockSize, numCols-iCol))
		{
			iRows =0;

			/* Loop over the block of rows */
			while  (iRows <  numRows )
			{
				D_Rprintf((" In third loop .. \n"));	
				vectorSize = min (rowBlockSize, numRows-iRows);

				sendToRank = currRow * numProcessCols + currCol;

				sendBuffer = gMatrix;

				startIndex = iRows + j * numRows + iCol * numRows ;

				/* 
				used for non-blocking communication

				PA_SendVectorToCR( &vectorSize, &one, sendBuffer+startIndex, &rowBlockSize, &sendToRank, requests+reqId);
				reqId++; 
				*/

				PA_SendVectorToCR( &vectorSize, &one, sendBuffer+startIndex, &rowBlockSize, &sendToRank);

				/*printContents (sendBuffer+startIndex, vectorSize);*/
				D_Rprintf(("Sent to rank - %d \n", sendToRank));

				currRow = (currRow+1) % numProcessRows;

				iRows = iRows + rowBlockSize;


			}

			j++;
			currRow = 0;	
		}

		currCol = (currCol+1) % numProcessCols;

		iCol = iCol + colBlockSize;	

	}

	D_Rprintf(("reqId : %d \n", reqId));	

	D_Rprintf(("Finished data distribution \n"));	

/* 
	used for non-blocking communication

	MPI_Waitall(reqId, requests, statuses); 
	free (requests);
	free (statuses);
*/	

}

void PAcollectData(double *retResult, int pgInfo[], int numRows, int numCols){

	int rowBlockSize, colBlockSize;
	int numProcessRows, numProcessCols;
	int currRow=0, currCol=0;
	int i=0,j=0;
	double *recvBuffer;
	int startIndex=0;
	int vectorSize;
	int one=1;
	int iColBlock=0, iRowBlock=0;
	int recvFromRank=0;
	int reqId=0;

/*	MPI_Request *requests;	
	MPI_Status *statuses;	
*/
	rowBlockSize = pgInfo[4];
	colBlockSize = pgInfo[5];
	numProcessRows = pgInfo[6];
	numProcessCols = pgInfo[7];
	recvBuffer = retResult;

	iColBlock = min( colBlockSize, numCols);


/* 
	used for non-blocking communication
	requests = (MPI_Request *) malloc (sizeof(MPI_Request) * rowBlockSize * iceil(numRows,rowBlockSize) * iceil(numCols,rowBlockSize));
	statuses = (MPI_Status *) malloc (sizeof(MPI_Status) * rowBlockSize * iceil(numRows,rowBlockSize) * iceil(numCols,rowBlockSize));
*/
	/* Handle the first block of column separately */
	while(i < iColBlock){
		currRow=0;

		iRowBlock = min(rowBlockSize, numRows);

		recvFromRank = currRow * numProcessCols + currCol;

		vectorSize = min (rowBlockSize, numRows);
		
		startIndex = i * numRows;

		/* 
		used for non-blocking communication

		PA_RecvVectorFromCR (&vectorSize, &one, recvBuffer+startIndex, &rowBlockSize, &recvFromRank, requests+reqId);
		reqId++;
		*/

		PA_RecvVectorFromCR (&vectorSize, &one, recvBuffer+startIndex, &rowBlockSize, &recvFromRank);
		currRow = ((currRow+1) % numProcessRows);

		/* Loop over the remainig block of rows */

		j =  iRowBlock;
		
		while ( j < numRows ){

			vectorSize = min (rowBlockSize, numRows-j);
			
			recvFromRank = currRow * numProcessCols + currCol;
			
			startIndex = i*numRows+j ;

			/* 
			used for non-blocking communication

			PA_RecvVectorFromCR (&vectorSize, &one, recvBuffer+startIndex, &rowBlockSize, &recvFromRank, requests+reqId);
			reqId++; 
			*/

			PA_RecvVectorFromCR (&vectorSize, &one, recvBuffer+startIndex, &rowBlockSize, &recvFromRank);

			j = j + rowBlockSize;

			currRow = ((currRow+1) % numProcessRows);

		}

		i++;
	}
			
	currCol = ((currCol+1) % numProcessCols);

	/* Handle the first block of column separately */

	while (iColBlock < numCols){

		iRowBlock = min(rowBlockSize,numRows);

		for (i=0;i < min (colBlockSize, numCols-iColBlock);i++) {
	
			currRow=0;

			vectorSize = min (rowBlockSize, numRows);

			recvFromRank = currRow * numProcessCols + currCol;
			
			startIndex = numRows*iColBlock + i*numRows ;
		
			/* 
			used for non-blocking communication

			PA_RecvVectorFromCR (&vectorSize, &one, recvBuffer+startIndex, &rowBlockSize, &recvFromRank, requests+reqId);
			reqId++; */

			PA_RecvVectorFromCR (&vectorSize, &one, recvBuffer+startIndex, &rowBlockSize, &recvFromRank);

			j = iRowBlock;

			currRow = ((currRow+1) % numProcessRows);

			/* Loop over the remainig block of rows */
			while ( j < numRows ){

				vectorSize = min (rowBlockSize, numRows-j);
				
				recvFromRank = currRow * numProcessCols + currCol;

				startIndex = numRows*iColBlock + i*numRows + j ;
				  
				/* 
				used for non-blocking communication

				PA_RecvVectorFromCR (&vectorSize, &one, recvBuffer+startIndex, &rowBlockSize, &recvFromRank, requests+reqId);
				reqId++;
				*/

				PA_RecvVectorFromCR (&vectorSize, &one, recvBuffer+startIndex, &rowBlockSize, &recvFromRank);

				j = j + rowBlockSize;

				currRow = ((currRow+1) % numProcessRows);
			}	


		}	
	
		currCol = ((currCol+1) % numProcessCols);
	
		iColBlock = iColBlock + colBlockSize;

	}
	D_Rprintf(("Finished data Collection  \n"));	

	/* 
	used for non-blocking communication

	MPI_Waitall(reqId, requests, statuses);
	
	free (requests);
	free (statuses);
	*/	
}
